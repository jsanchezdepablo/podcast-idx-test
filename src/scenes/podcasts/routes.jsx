import React, { lazy } from "react";
import { Route } from "react-router-dom";
import SuspenseWrapper from "components/suspense-wrapper";

const PodcastsView = lazy(() => import("."));
const DetailView = lazy(() => import("./components/detail"));
const EpisodeView = lazy(() => import("./components/episode"));

const PODCASTS = {
  PATH: "/",
  DETAIL_PATH: "/podcast/:id",
  EPISODE_PATH: "/podcast/:id/episode/:id",
};

const podcastsRoutes = [
  <Route key="detail" exact path={PODCASTS.DETAIL_PATH}>
    {SuspenseWrapper(DetailView)}
  </Route>,
  <Route key="episode" exact path={PODCASTS.EPISODE_PATH}>
    {SuspenseWrapper(EpisodeView)}
  </Route>,
  <Route key="podcasts" exact path={PODCASTS.PATH}>
    {SuspenseWrapper(PodcastsView)}
  </Route>,
];

export default podcastsRoutes;
