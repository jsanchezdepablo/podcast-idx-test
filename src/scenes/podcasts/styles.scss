@import "assets/styles/colors.scss";

.podcast {
  &__filter-container {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    gap: 10px;
  }

  &__filter-count {
    padding: 0 4px 2px 4px;
    background-color: $primary-color;
    border-radius: 7px;
    font-weight: 500;
    color: white;
  }

  &__cards {
    margin-top: 80px;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
  }
}
