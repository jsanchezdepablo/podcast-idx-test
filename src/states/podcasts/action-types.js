const ACTION_TYPES = {
  SET_PODCASTS: "SET_PODCASTS",
  SET_PODCAST: "SET_PODCAST",
  SET_IS_LOADING: "SET_IS_LOADING",
  RESET: "RESET",
};

export default ACTION_TYPES;
