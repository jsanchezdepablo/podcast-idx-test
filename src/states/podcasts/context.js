import React, { createContext, useContext, useReducer, useMemo } from "react";
import Reducer from "./reducer";
import INIT_STATE from "./init-state";

const Context = createContext();
export const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(Reducer, INIT_STATE);
  const contextValue = useMemo(() => ({ state, dispatch }), [state, dispatch]);
  return <Context.Provider value={contextValue}>{children}</Context.Provider>;
};
export const Selector = () => useContext(Context);
