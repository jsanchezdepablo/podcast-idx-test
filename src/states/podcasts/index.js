import { Provider, Selector } from "./context";
import Actions from "./actions";

export const PodcastsProvider = Provider;
export const usePodcastsSelector = Selector;
export const PodcastsActions = Actions;
