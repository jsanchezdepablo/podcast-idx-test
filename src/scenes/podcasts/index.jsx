import React, { useEffect, useMemo } from "react";
import TextField from "@mui/material/TextField";
import { usePodcastsSelector, PodcastsActions } from "states/podcasts";
import DefaultCircularProgress from "components/circular-progress";
import Card from "./components/card";

import "./styles.scss";

const PodcastsView = () => {
  const {
    state: { podcasts, isLoading },
  } = usePodcastsSelector();

  const { searchPodcasts } = PodcastsActions.useFetchPodcasts();

  useEffect(() => {
    searchPodcasts();
  }, []);

  const podcastCards = useMemo(() => {
    if (podcasts.length > 0) {
      return podcasts.map((podcast) => <Card key={podcast.title} data={podcast} />);
    }
  }, [podcasts]);

  return (
    <div>
      <div className="podcast__filter-container">
        <div className="podcast__filter-count">{podcasts.length}</div>
        <TextField label="Filter podcasts" variant="outlined" size="small" />
      </div>
      {isLoading ? <DefaultCircularProgress /> : <div className="podcast__cards">{podcastCards}</div>}
    </div>
  );
};

export default PodcastsView;
