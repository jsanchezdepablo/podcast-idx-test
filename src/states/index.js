import React from "react";
import { PodcastsProvider } from "./podcasts";

const GlobalStateProvider = ({ children }) => <PodcastsProvider>{children}</PodcastsProvider>;

export default GlobalStateProvider;
