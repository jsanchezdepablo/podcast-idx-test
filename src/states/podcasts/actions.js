import { useEffect } from "react";
import { useFetchApi } from "states/utils";
import { Selector } from "./context";
import ACTION_TYPES from "./action-types";

const URL_PODCASTS = "https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json";

export const useFetchPodcasts = () => {
  const { dispatch } = Selector();
  const { search: searchApi, result, isLoading } = useFetchApi(URL_PODCASTS);

  useEffect(() => {
    if (result != null) {
      dispatch({ type: ACTION_TYPES.SET_PODCASTS, payload: result });
    }
  }, [result]);

  const searchPodcasts = () => {
    if (!isLoading) {
      searchApi();
    }
  };

  return { searchPodcasts };
};

const actions = { useFetchPodcasts };

export default actions;
