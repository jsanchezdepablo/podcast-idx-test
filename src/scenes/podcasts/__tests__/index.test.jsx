import React from "react";
import { shallow } from "enzyme";
import * as PodcastsState from "states/podcasts";
import PodcastsView from "../";

const MOCK_PODCASTS = [
  {
    id: "1",
    title: "title-1",
    artist: "artist-1",
    image: "//image",
    description: "description",
  },
  {
    id: "2",
    title: "title-2",
    artist: "artist-2",
    image: "//image",
    description: "description",
  },
];
describe("PodcastsView", () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetAllMocks();

    jest
      .spyOn(PodcastsState, "usePodcastsSelector")
      .mockImplementation(() => ({ state: { podcasts: [], isLoading: false, lastUpdatedDate: "date" } }));

    jest
      .spyOn(PodcastsState.PodcastsActions, "useFetchPodcasts")
      .mockImplementation(() => ({ searchPodcasts: () => {} }));
  });

  it("Matches snapshot", () => {
    const wrapper = shallow(<PodcastsView />);
    expect(wrapper).toMatchSnapshot();
  });

  it("Should render loader", () => {
    jest.spyOn(PodcastsState, "usePodcastsSelector").mockImplementation(() => ({ state: { isLoading: true } }));

    const wrapper = shallow(<PodcastsView />);
    expect(wrapper.find("DefaultCircularProgress").exists()).toBeTruthy();
  });

  it("Should render podcast cards", () => {
    jest
      .spyOn(PodcastsState, "usePodcastsSelector")
      .mockImplementation(() => ({ state: { podcasts: MOCK_PODCASTS, isLoading: false } }));

    const wrapper = shallow(<PodcastsView />);
    expect(wrapper.find(".podcast__cards").exists()).toBeTruthy();
  });

  it("Should call onChange Textfield", () => {
    const wrapper = shallow(<PodcastsView />);
    wrapper
      .find("ForwardRef(TextField)")
      .props()
      .onChange({ target: { value: "value" } });
  });
});
