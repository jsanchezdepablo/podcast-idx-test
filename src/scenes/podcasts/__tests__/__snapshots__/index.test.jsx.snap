// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodcastsView Matches snapshot 1`] = `
<div>
  <div
    className="podcast__filter-container"
  >
    <div
      className="podcast__filter-count"
    >
      0
    </div>
    <ForwardRef(TextField)
      label="Filter podcasts"
      onChange={[Function]}
      size="small"
      variant="outlined"
    />
  </div>
  <div
    className="podcast__cards"
  />
</div>
`;
